### PR TITLE
ci(pipeline): add release/deploy buttons

### PR DESCRIPTION
**Pull request by lionelB**
this PR 
- replace `ENVIRONMENT` by `DOMAIN`
- add deploy button in pipeline that will run trigger with production=true
- reuse master's monorepo image if there no `package.json` update 
- reuse monorepo image when release a version
- skip stages prepare, quality, register when deploy in production,

### CODE REVIEW

File: .gitlab-ci/stages/notify.yml [hunk 1]
```diff
@@ -45,12 +47,15 @@ Notify Success:
...
     - sh ./.k8s/scripts/send-url.sh $(cat $DEPLOY_ID_FILE) success
 
+#
+#
+#
+
 Release:
   stage: "Notify Finished Deployment"
-  image: node:10
+  image: $CI_REGISTRY_IMAGE:$CI_COMMIT_SHA
```
Comment (revolunet) [inline .gitlab-ci/stages/notify.yml:56]: pourquoi pas  node:10 ? (ou 12)

File: .gitlab-ci/stages/prepare.yml [hunk 2]
```diff
@@ -4,16 +4,52 @@
...
+      fi
+    #
+    - docker pull $IMAGE_NAME:master || true
+    - echo "> docker tag ${IMAGE_NAME}:master ${IMAGE_NAME}:${TAG}"
+    - docker tag $IMAGE_NAME:master $IMAGE_NAME:$CI_COMMIT_SHA
+    - echo "> docker tag ${IMAGE_NAME}:master ${IMAGE_NAME}:${CI_COMMIT_SHA}"
+    - docker tag $IMAGE_NAME:master $IMAGE_NAME:$TAG
+    - docker push $IMAGE_NAME:$CI_COMMIT_SHA
+    - docker push $IMAGE_NAME:$TAG
+#
```
Comment (revolunet) [inline .gitlab-ci/stages/prepare.yml:55]: si je comprends bien quand il n'y a pas de changement à l'image de base (basé sur les deps) tu repush la master avec un nouveau tag+hash ? (meme image)

File: .gitlab-ci/stages/notify.yml (hunk 1 above)
Comment (lionelB) [reply .gitlab-ci/stages/notify.yml:56]: car on reutilise l'image du mono-repo avec les node_modules, comme ca pas besoin de `yarn --frozen-lockfile` on peut directement lancer `lerna version`

File: .gitlab-ci/stages/prepare.yml (hunk 2 above)
Comment (lionelB) [reply .gitlab-ci/stages/prepare.yml:55]: yep

Comment (revolunet) [reply .gitlab-ci/stages/prepare.yml:55]: ok, nice trick

**Review by revolunet — APPROVED** [review]

**Comment by revolunet** [comment]
Could we use the same only/except clauses to only deploy frontend when only frontend/ui/old-ui have changed ?

**Comment by lionelB** [comment]
> Could we use the same only/except clauses to only deploy frontend when only frontend/ui/old-ui have changed ?

Yup! this is the idea, but register images (api / frontend) have more dependencies

File: .gitlab-ci/stages/notify.yml [hunk 3]
```diff
@@ -61,6 +66,26 @@ Release:
...
     - GH_TOKEN=${GITHUB_TOKEN} yarn lerna version ${LERNA_ARGS:="--yes"}
 
+#
+#
+#
+
+Production:
+  <<: *notify_stage
+  stage: "Notify Finished Deployment"
+  image: $CI_REGISTRY_IMAGE:$CI_COMMIT_SHA
```
Comment (douglasduteil) [inline .gitlab-ci/stages/notify.yml]: we actually just need `curl` here

File: .gitlab-ci/stages/prepare.yml [hunk 4]
```diff
@@ -4,16 +4,52 @@
...
+Register new socialgouv/cdtn base image:
+  <<: *prepare_stage
+  variables:
+    CONTEXT: .
+    IMAGE_NAME: $CI_REGISTRY_IMAGE
+  only:
+    changes:
+      - yarn.lock
+      - package.lock
+      - packages/**/package.lock
```
Comment (douglasduteil) [inline .gitlab-ci/stages/prepare.yml:19]: The base image contains build files too no ?

**Review by douglasduteil — APPROVED** [review]

File: .gitlab-ci/stages/notify.yml (hunk 3 above)
Comment (lionelB) [reply .gitlab-ci/stages/notify.yml]: yes indeed ! we coud use a lighter image

File: .gitlab-ci/stages/prepare.yml (hunk 4 above)
Comment (lionelB) [reply .gitlab-ci/stages/prepare.yml:19]: only for data but you're right we need to add these file as a dependency

**Review by revolunet — APPROVED** [review]

**Comment by revolunet** [comment]
lets try this !!

**Pull request merged**